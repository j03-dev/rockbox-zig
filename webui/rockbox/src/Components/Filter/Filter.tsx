/* eslint-disable @typescript-eslint/no-explicit-any */
import { Theme, useTheme } from "@emotion/react";
import { Input } from "baseui/input";
import { FC, useEffect } from "react";
import { Controller, useForm } from "react-hook-form";
import Search from "../Icons/Search";

export type FilterProps = {
  placeholder?: string;
  onChange: (value: string) => void;
};

const Filter: FC<FilterProps> = ({ placeholder = "Filter", onChange }) => {
  const theme = useTheme();
  const { control, watch } = useForm({
    defaultValues: {
      filter: "",
    },
  });

  useEffect(() => {
    const subscription = watch((value, { name, type }) => {
      if (type === "change") {
        onChange(value[name!] as string);
      }
    });
    return () => subscription.unsubscribe();
  }, [onChange, watch]);

  return (
    <>
      <Controller
        render={({ field }) => (
          <Input
            {...(field as any)}
            startEnhancer={<Search />}
            placeholder={placeholder}
            overrides={styles.filter(theme)}
          />
        )}
        control={control}
        name="filter"
        rules={{ required: true }}
      />
    </>
  );
};

const styles = {
  filter: (theme: Theme) => ({
    Root: {
      style: {
        height: "40px",
        width: "222px",
        borderTopWidth: "1.5px !important",
        borderBottomWidth: "1.5px !important",
        borderLeftWidth: "1.5px !important",
        borderRightWidth: "1.5px !important",
        borderTopColor: "rgba(82, 82, 82, 0.0625) !important",
        borderBottomColor: "rgba(82, 82, 82, 0.0625) !important",
        borderLeftColor: "rgba(82, 82, 82, 0.0625) !important",
        borderRightColor: "rgba(82, 82, 82, 0.0625) !important",
        borderTopLeftRadius: "20px !important",
        borderTopRightRadius: "20px !important",
        borderBottomLeftRadius: "20px !important",
        borderBottomRightRadius: "20px !important",
        backgroundColor: theme.colors.searchBackgroundAlt,
      },
    },
    Input: {
      style: {
        backgroundColor: theme.colors.searchBackgroundAlt,
        fontSize: "14px",
        color: theme.colors.text,
      },
    },
    InputContainer: {
      style: {
        backgroundColor: theme.colors.searchBackgroundAlt,
      },
    },
    StartEnhancer: {
      style: {
        backgroundColor: theme.colors.searchBackgroundAlt,
        paddingTop: "0px",
        paddingBottom: "0px",
        paddingLeft: "0px",
        paddingRight: "0px",
      },
    },
  }),
};

export default Filter;
