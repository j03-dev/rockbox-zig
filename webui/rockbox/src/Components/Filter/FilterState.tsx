import { atom } from "recoil";
import { SearchResults } from "../../Hooks/GraphQL";

export const filterState = atom<{
  term: string;
  results?: SearchResults;
}>({
  key: "filterState",
  default: {
    term: "",
    results: undefined,
  },
});
