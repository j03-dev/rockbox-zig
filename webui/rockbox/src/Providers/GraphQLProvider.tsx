import {
  ApolloClient,
  createHttpLink,
  InMemoryCache,
  ApolloProvider,
  split,
  from,
} from "@apollo/client";
import { removeTypenameFromVariables } from "@apollo/client/link/remove-typename";
import { WebSocketLink } from "@apollo/client/link/ws";
import { getMainDefinition } from "@apollo/client/utilities";
import { SubscriptionClient } from "subscriptions-transport-ws";
import { FC, ReactNode } from "react";

const uri =
  process.env.NODE_ENV === "development"
    ? import.meta.env.VITE_APP_API_URL || "http://localhost:6062/graphql"
    : `${origin}/graphql`;

const removeTypenameLink = removeTypenameFromVariables();

const httpLink = createHttpLink({
  uri,
});

const wsLink = new WebSocketLink(
  new SubscriptionClient(uri.replace("http", "ws"))
);

const splitLink = split(
  ({ query }) => {
    const definition = getMainDefinition(query);
    return (
      definition.kind === "OperationDefinition" &&
      definition.operation === "subscription"
    );
  },
  wsLink,
  from([removeTypenameLink, httpLink])
);
const link = splitLink;

const client = new ApolloClient({
  link,
  cache: new InMemoryCache(),
});

const GraphQLProvider: FC<{
  children: ReactNode;
}> = ({ children }) => {
  return <ApolloProvider client={client}>{children}</ApolloProvider>;
};

export default GraphQLProvider;
