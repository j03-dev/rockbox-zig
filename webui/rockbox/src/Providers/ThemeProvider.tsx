import { createContext, useState, FC } from "react";
import { ThemeProvider as EmotionThemeProvider } from "@emotion/react";
import {
  BaseUIDarkTheme,
  BaseUILightTheme,
  DarkTheme,
  LightTheme,
} from "../Theme";
import { BaseProvider } from "baseui";
import { ThemeProvider as MaterialThemeProvider } from "@mui/material/styles";
import { createTheme } from "@mui/material/styles";

export type Theme = "light" | "dark";

export type ThemeContextType = {
  theme: Theme;
  setTheme: (theme: Theme) => void;
};

export const ThemeContext = createContext<ThemeContextType>({
  theme: "light",
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  setTheme: (_theme: Theme) => {},
});

const muitheme = createTheme({
  cssVariables: true,
  palette: {
    primary: {
      main: "#fe099c",
    },
  },
});

export type ThemeProviderProps = {
  children: React.ReactNode;
};

const ThemeProvider: FC<ThemeProviderProps> = ({ children }) => {
  const [theme, setTheme] = useState<Theme>("light");
  return (
    <MaterialThemeProvider theme={muitheme}>
      <ThemeContext.Provider value={{ theme, setTheme }}>
        <EmotionThemeProvider theme={theme === "dark" ? DarkTheme : LightTheme}>
          <BaseProvider
            theme={theme === "dark" ? BaseUIDarkTheme : BaseUILightTheme}
          >
            {children}
          </BaseProvider>
        </EmotionThemeProvider>
      </ThemeContext.Provider>
    </MaterialThemeProvider>
  );
};

export default ThemeProvider;
