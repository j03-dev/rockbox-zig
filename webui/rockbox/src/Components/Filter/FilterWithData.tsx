import { FC } from "react";
import Filter from "./Filter";
import { useSearchLazyQuery } from "../../Hooks/GraphQL";
import { useRecoilState } from "recoil";
import { filterState } from "./FilterState";
import _ from "lodash";

const FilterWithData: FC<{ placeholder?: string }> = (props) => {
  const [, setFilterState] = useRecoilState(filterState);
  const [search] = useSearchLazyQuery();
  const onSearch = async (term: string) => {
    setFilterState((state) => ({ ...state, term }));
    if (term.length > 2) {
      _.debounce(async () => {
        const { data } = await search({ variables: { term } });
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        setFilterState((state) => ({ ...state, results: data?.search as any }));
      }, 500)();
    }
    if (term.length === 0) {
      setFilterState({ term: "", results: undefined });
    }
  };
  return <Filter onChange={onSearch} {...props} />;
};

export default FilterWithData;
