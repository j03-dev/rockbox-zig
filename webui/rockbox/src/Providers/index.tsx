import { FC } from "react";
import ThemeProvider from "./ThemeProvider";
import { Provider as StyletronProvider } from "styletron-react";
import { Client as Styletron } from "styletron-engine-atomic";
import { RecoilRoot } from "recoil";

const engine = new Styletron();

export type ProvidersProps = {
  children: React.ReactNode;
};

const Providers: FC<ProvidersProps> = ({ children }) => {
  return (
    <RecoilRoot>
      <StyletronProvider value={engine}>
        <ThemeProvider>{children}</ThemeProvider>
      </StyletronProvider>
    </RecoilRoot>
  );
};

export default Providers;
