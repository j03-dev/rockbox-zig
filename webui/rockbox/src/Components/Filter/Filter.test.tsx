import { render } from "@testing-library/react";
import { vi } from "vitest";
import Filter from "./Filter";
import Providers from "../../Providers";

describe("Filter", () => {
  it("should render", () => {
    const { container } = render(
      <Providers>
        <Filter placeholder="Search Songs" onChange={vi.fn()} />
      </Providers>
    );
    expect(container).toMatchSnapshot();
  });
});
