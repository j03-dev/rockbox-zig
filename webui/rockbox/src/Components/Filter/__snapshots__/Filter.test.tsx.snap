// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Filter > should render 1`] = `
<div>
  <div
    class=""
  >
    <div
      class="ae af ag ah ai aj ak al am an ao ap aq ar as at au av aw ax ay az b0 b1 b2 b3 b4 b5 b6 b7 b8"
      data-baseweb="input"
    >
      <div
        class="af b9 ba bb ar as ax ay az b0 b7 bc bd b5 be bf"
      >
        <svg
          data-test="icon-ActionActiveSearch"
          fill="none"
          height="20"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M4 9.5a5.5 5.5 0 1 1 11 0 5.5 5.5 0 0 1-11 0ZM9.5 2a7.5 7.5 0 1 0 4.549 13.463l2.744 2.744a1 1 0 0 0 1.414-1.414l-2.744-2.744A7.5 7.5 0 0 0 9.5 2Z"
            fill="#ACAAAA"
            fill-rule="evenodd"
          />
        </svg>
      </div>
      <div
        class="af bg bh ar as ax ay az b0 bd b5"
        data-baseweb="base-input"
      >
        <input
          aria-invalid="false"
          aria-required="false"
          autocomplete="on"
          class="ae b5 bi bj bk bl bm bn bo bp bq bg br bs bt bu bv bw b6 bx ax by az b0 bz c0 c1"
          inputmode="text"
          name="filter"
          placeholder="Search Songs"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
  <div
    class=""
  />
</div>
`;
